src/pymoca/tree.py

```diff
@@ -259,6 +273,694 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+
+    # If not found and we stopped at an encapsulated class,
+    # then search predefined functions and operators in global scope
+    # TODO: Add predefined functions and operators to global scope before this
+    if found is None and current_scope.encapsulated:
+        found = _find_local(name, scope.root)
+        if not isinstance(found, ast.Class) or found.type not in ("function", "operator"):
+            found = None
+
+    return found
+
+
+def _find_rest_of_name(
+    first: Union[ast.Class, ast.Symbol], rest_of_name: str
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Lookup the `B.C` part of Composite Name Lookup (`A.B.C`) (spec 5.3.2)
+
+    1. `A` is looked up using Simple Name Lookup and passed as `first` argument
+    2. If `A` is a Component:
+        1. `B.C` is looked up from named component elements of `A`
+        2. if not found and if `A.B.C` is used as a function call and `A` is a scalar or can be
+        evaluated as a scalar from an array and `B` and `C` are classes,
+        it is a non-operator function call.
+    3. If `A` is a Class:
+        1. `A` is temporarily flattened without modifiers
+        2. `B.C` is looked up among named elements of temp flattened class,
+        but if `A` is not a package, lookup is restricted to `encapsulated` elements only
+        and "the class we look inside shall not be partial in a simulation model"."""
+    if isinstance(first, ast.Symbol):
+        # Find the symbol type
+        if isinstance(first.type, ast.Class):
+            type_class = first.type
+        else:
+            type_class = find_name(first.type, first.parent, copy=False)
+            if type_class is None:
+                full_ref = str(first.parent.full_reference()) + "." + first.name
+                raise NameLookupError(f"Lookup failed for type of symbol {full_ref}")
+        found = _find_composite_name_in_symbols(rest_of_name, type_class)
+        if not found:
+            # Can only find in classes if the below rules apply:
+            # 2b. if not found and if `A.B.C` is used as a function call
+            # and `A` is a scalar or can be evaluated as a scalar from an array
+            # and `B` and `C` are classes,
+            # it is a non-operator function call.
+            found = _find_composite_name_in_classes(rest_of_name, type_class)
+            if isinstance(found, ast.Class):
+                if found.type != "function":
+                    found = None
+                else:
+                    if first.dimensions[0][0].value is not None:
+                        raise NameLookupError(
+                            f"Array {first.name} must have subscripts to lookup function {found.name}"
+                        )
+
+    elif isinstance(first, ast.Class):
+        found = _flatten_first_and_find_rest(first, rest_of_name)
+    else:
+        raise NameLookupError(f'Found unexpected node "{found!r}" during name lookup')
```

Comment:
Yep, that should be `first`, not `found`. Made a commit to fix. 

Code coverage is on this PR TODO list after flattening because I may yet uncover some name lookup refactoring needed.